### PR TITLE
Version 4.3.7

### LINKED ISSUE #2665: Canceling async Redis command leaves connection open, in unsafe state for future commands
**Version**: 4.5.3

**Platform**: Python 3.8 on Ubuntu  / Generic

**Description**: Canceling async Redis command leaves connection open, in unsafe state for future commands

This is a reincarnation of #2624, which was closed with an incomplete fix and a possibly unreliable test case. This is the same issue that recently got a lot of attention due to ChatGPT outage, and that remains only partially fixed. The cancellation shielding introduced in #2641 addressed only the cancellation of Redis pipeline operation, but non-pipelined ops are still vulnerable.

This time I am attaching a script that reproduces the issue reliably without relying on an external, slow Redis server. This is achieved by inserting a small TCP socket proxy between the Redis client and local Redis server, with the proxy introducing a 0.1 second delay when sending data in either direction. 

Running this script with a Redis server running locally on port 6379 produces the following output:
```
$ python redis_cancel.py 
managed to cancel the task, connection is left open with unread response
bar: b'foo'
ping: False
foo: b'PONG'
```

```python
import asyncio

from redis.asyncio import Redis


async def pipe(reader: asyncio.StreamReader, writer: asyncio.StreamWriter, delay: float, name=''):
    while data := await reader.read(1000):
        # print(name, 'received:', data)
        await asyncio.sleep(delay)
        writer.write(data)
        await writer.drain()


class DelayProxy:

    def __init__(self, addr, redis_addr, delay: float):
        self.addr = addr
        self.redis_addr = redis_addr
        self.delay = delay

    async def start(self):
        server = await asyncio.start_server(self.handle, *self.addr)
        asyncio.create_task(server.serve_forever())

    async def handle(self, reader, writer):
        # establish connection to redis
        redis_reader, redis_writer = await asyncio.open_connection(*self.redis_addr)
        pipe1 = asyncio.create_task(pipe(reader, redis_writer, self.delay, 'to redis:'))
        pipe2 = asyncio.create_task(pipe(redis_reader, writer, self.delay, 'from redis:'))
        await asyncio.gather(pipe1, pipe2)


async def main():

    # create a tcp socket proxy that relays data to Redis and back, inserting 0.1 seconds of delay
    dp = DelayProxy(addr=('localhost', 6380), redis_addr=('localhost', 6379), delay=0.1)
    await dp.start()

    # note that we connect to proxy, rather than to Redis directly
    async with Redis(host='localhost', port=6380) as r:

        await r.set('foo', 'foo')
        await r.set('bar', 'bar')

        t = asyncio.create_task(r.get('foo'))
        await asyncio.sleep(0.050)
        t.cancel()
        try:
            await t
            print('try again, we did not cancel the task in time')
        except asyncio.CancelledError:
            print('managed to cancel the task, connection is left open with unread response')

        print('bar:', await r.get('bar'))
        print('ping:', await r.ping())
        print('foo:', await r.get('foo'))

if __name__ == '__main__':
    asyncio.run(main())
```

### PR DESCRIPTION
**Pull request by dvora-h**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [ ] Does `$ tox` pass with this change (including linting)?
- [ ] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [ ] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Is there an example added to the examples folder (if applicable)?
- [ ] Was the change added to CHANGES file?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change

Closes #2665

### CODE REVIEW

File: tests/test_asyncio/test_cwe_404.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+class DelayProxy:
+    def __init__(self, addr, redis_addr, delay: float):
+        self.addr = addr
+        self.redis_addr = redis_addr
+        self.delay = delay
+
+    async def start(self):
+        self.server = await asyncio.start_server(self.handle, *self.addr)
+        self.ROUTINE = asyncio.create_task(self.server.serve_forever())
```
Comment (jamestiotio) [inline tests/test_asyncio/test_cwe_404.py:30]:
```suggestion
        self.ROUTINE = asyncio.ensure_future(self.server.serve_forever())
```

File: tests/test_asyncio/test_cwe_404.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    async def handle(self, reader, writer):
+        # establish connection to redis
+        redis_reader, redis_writer = await asyncio.open_connection(*self.redis_addr)
+        pipe1 = asyncio.create_task(pipe(reader, redis_writer, self.delay, "to redis:"))
```
Comment (jamestiotio) [inline tests/test_asyncio/test_cwe_404.py:35]:
```suggestion
        pipe1 = asyncio.ensure_future(pipe(reader, redis_writer, self.delay, "to redis:"))
```

File: tests/test_asyncio/test_cwe_404.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        pipe2 = asyncio.create_task(
```
Comment (jamestiotio) [inline tests/test_asyncio/test_cwe_404.py:36]:
```suggestion
        pipe2 = asyncio.ensure_future(
```

File: tests/test_asyncio/test_cwe_404.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    await dp.start()
+
+    for b in [True, False]:
+        # note that we connect to proxy, rather than to Redis directly
+        async with Redis(host="localhost", port=5380, single_connection_client=b) as r:
+
+            await r.set("foo", "foo")
+            await r.set("bar", "bar")
+
+            t = asyncio.create_task(r.get("foo"))
```
Comment (jamestiotio) [inline tests/test_asyncio/test_cwe_404.py:66]:
```suggestion
            t = asyncio.ensure_future(r.get("foo"))
```

File: tests/test_asyncio/test_cwe_404.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        await r.set("bar", "bar")
+
+        pipe = r.pipeline()
+
+        pipe2 = r.pipeline()
+        pipe2.get("bar")
+        pipe2.ping()
+        pipe2.get("foo")
+
+        t = asyncio.create_task(pipe.get("foo").execute())
```
Comment (jamestiotio) [inline tests/test_asyncio/test_cwe_404.py:103]:
```suggestion
        t = asyncio.ensure_future(pipe.get("foo").execute())
```

File: tests/test_asyncio/test_cwe_404.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+    dp = DelayProxy(addr=("localhost", 5381), redis_addr=("localhost", 6372), delay=0.1)
+    await dp.start()
+
+    r = RedisCluster.from_url("redis://localhost:5381")
+    await r.initialize()
+    await r.set("foo", "foo")
+    await r.set("bar", "bar")
+
+    t = asyncio.create_task(r.get("foo"))
```
Comment (jamestiotio) [inline tests/test_asyncio/test_cwe_404.py:136]:
```suggestion
    t = asyncio.ensure_future(r.get("foo"))
```

**Review by jamestiotio — CHANGES_REQUESTED** [review]
For Python 3.6, these calls to `asyncio.create_task` should be modified to `asyncio.ensure_future` instead.

EDIT: Seems like `asyncio.CancelledError` existed in [Python 3.6.0's code](https://github.com/python/cpython/blob/v3.6.0/Lib/asyncio/base_futures.py#L9) after all, it was just undocumented (earliest documented page of [asyncio exceptions](https://docs.python.org/3.7/library/asyncio-exceptions.html) was for Python 3.7). Ignore my previous comment about `asyncio.CancelledError`.

**Comment by auvipy** [comment]
will be looking forward to this release

**Comment by chayim** [comment]
Using the proxy changes from @drago-balto, combined with the current 4.3 branch yields a sideaffect in this particular case. Given the output errors on RuntimeError - we can now be positive, that the fix is truly caught.

However, the byproduct of that is this given case. If you cancel an executing pipeline (as [in our test](https://github.com/redis/redis-py/pull/2672/files#diff-a90c3a19dad7803c9726358f223a3b7b8fb23ccb1d573d580d6640f04fdd3f27R103) then the pipeline object cannot be reused.

Technically, it's always been possible (and works) to reuse the pipeline object, rather than reinitialize it:

```
r = Redis()
pipe = r.pipeline()
pipe.set("foo", "bar")
pipe.execute()

pipe.set("bar", "blee")
pipe.get("foo")
pipe.execute()
````

However, in 4.3 this won't be possible *should* the async be canceled. IMHO this is reasonable, it's a security fix we're backporting to a dead branch. Meaning if you end up in this situation, you can easily catch RuntimeError and move along.

Thoughts community?

**Comment by chayim** [comment]
@auvipy I haven't seen feedback from anyone.. I let it bake for two weeks, to make sure there's real interest here. Can you confirm - is there - and does this fix work happily for you?  Anyone else, please feel free to weigh in.

This version is old enough, that I want to double-check here.

**Review by auvipy — COMMENTED** [review]
at celery we use mostly the sync versions codes so far

**Comment by auvipy** [comment]
but we are py 3.7+ and pypy 3.8+

**Comment by chayim** [comment]
Right. Since celery is py 3.7+ this PR isn't needed. That's already covered by existing releases.

@dvora-h Let's let this bake. If no one *needs* the change, specifically due to python 3.6, we abandon.

**Comment by auvipy** [comment]
but we need to calm down pyup

**Comment by chayim** [comment]
> but we need to calm down pyup

That sounds like a misconfiguration of pyup and IMHO not a reason to release this. I'd rather not extend a potential issue.

Celery releases clearly support Python 3.7+, especially since 3.6 has been end of lifed.  I went back through the history of the repository and Pypi, and haven't seen an active version support Python 3.6.

This has sat long enough, I think we can close. Thank you everyone who helps us get to this point!

**Pull request closed without merging**